### DISCUSSION TITLE
Strange behaviour of pick_fastest() in Imola Qualifying

### DISCUSSION THREAD

**Discussion by blakehinsey** [theOehrly/Fast-F1#164] [General]
Has anyone noticed the behaviour of the pick_fastest() with Imola data for Qualifying?

I've looked into the code, and the only_by_time=True helps.

![image](https://user-images.githubusercontent.com/20929863/165350151-ff3dd698-ca6a-4c55-a56f-124ec55d590c.png)

Does this mean that Q1 and Q2 laps were not counted as personal bests? It is a strange one as Q1/2 were dry and Q3 times were wet.

I will admit, my coding is not excellent, so maybe I have goofed something up.

It's "BRRRAKE" from Twitter btw!

**Comment by theOehrly** [comment]
I see the problem. It looks like the API resets personal best for every quali section. It makes sense when watching quali timing live. But I only keep a reference to the last received personal best time in the API parser. Therefore, Q3 overrides Q2 here despite the times being slower. This could actually happen in some dry sessions too, which is not good at all.

Working around this is non-trivial. I'd need to keep track of the order of personal best times to kind of create an official ranking. While still detecting lap time deletions. Additionally, it needs to be possible to split the Laps into Q1, Q2 and Q3 laps. It's possible, but somewhat more complicated than the current implementation.

**Reply by blakehinsey** [reply]
I think getting the best Q1/2/3 from the SessionResults is straight forward at least in terms of laptime and classification, but is it possible to use that object to access the associated telemetry?

**Reply by theOehrly** [reply]
No, you cannot use that. You need to get a reference to the associated lap. I guess, you could get it by finding the lap with the same lap time.  Then you can access the telemetry from there.

I guess that would be a nice feature too, actually. Getting lap objects from session results.
